Spigot-Server-Patches/0696-Skip-unnecessary-hopper-pulls-and-pushes.patch

```diff
@@ -0,0 +1,279 @@
+ 
+-                if (!this.isEmpty()) {
++                if (!shouldSkipPush() && !this.isEmpty()) { // Paper - don't attempt a futile push
+                     flag = this.k();
+                 }
++                lastAttemptedPush = MinecraftServer.currentTick; // Paper
+ 
+-                if (!this.j()) {
++                if (!shouldSkipPull() && !this.j()) { // Paper - don't attempt a futile pull
+                     flag |= (Boolean) supplier.get();
+                 }
++                lastAttemptedPull = MinecraftServer.currentTick; // Paper
+ 
+                 if (flag) {
+                     this.setCooldown(world.spigotConfig.hopperTransfer); // Spigot
+@@ -221,6 +225,7 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+                     itemstack = callPushMoveEvent(iinventory, itemstack);
+                     if (itemstack == null) { // cancelled
+                         origItemStack.setCount(origCount);
++                        lastSuccessfulPush = MinecraftServer.currentTick; // Count as success because plugins are unpredictable
+                         return false;
+                     }
+                 }
+@@ -234,6 +239,7 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+                     }
+                     this.setItem(i, origItemStack);
+                     iinventory.update();
++                    lastSuccessfulPush = MinecraftServer.currentTick;
+                     return true;
+                 }
+                 origItemStack.setCount(origCount);
+@@ -256,6 +262,7 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+             itemstack = callPullMoveEvent(ihopper, iinventory, itemstack);
+             if (itemstack == null) { // cancelled
+                 origItemStack.setCount(origCount);
++                if (ihopper instanceof TileEntityHopper) ((TileEntityHopper) ihopper).lastSuccessfulPull = MinecraftServer.currentTick; // Count as a success because plugins are unpredictable
+                 // Drastically improve performance by returning true.
+                 // No plugin could of relied on the behavior of false as the other call
+                 // site for IMIE did not exhibit the same behavior
+@@ -275,6 +282,7 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+             iinventory.setItem(i, origItemStack);
+             IGNORE_TILE_UPDATES = false;
+             iinventory.update();
++            if (ihopper instanceof TileEntityHopper) ((TileEntityHopper) ihopper).lastSuccessfulPull = MinecraftServer.currentTick;
+             return true;
+         }
+         origItemStack.setCount(origCount);
+@@ -349,6 +357,81 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+         }
+     }
+     // Paper end
++    // Paper start - skip unnecessary hopper pulls and pushes
```

Comment:
Don't think we need to do this, we can just continue the last paper start but could be wrong

Reply:
I wasn't sure about this one either - thought it might be better to have a separate block to be clear which patch they belong to. Maybe someone else can give their input on this

Reply:
Whatever you feel is best, really. It really doesn't matter much as most people use something like [GitToolBox](https://plugins.jetbrains.com/plugin/7499-gittoolbox)'s inline blame anyways.

Reply:
Okay, cool. I think I'll just keep it as it is now